File: app/src/main/java/com/spiderbiggen/randomchampionselector/views/TopCenteredImageView.java
package com.spiderbiggen.randomchampionselector.views;

import android.content.Context;
import android.graphics.Matrix;
import android.graphics.drawable.Drawable;
import android.support.annotation.Nullable;
import android.util.AttributeSet;

/**
 * Created on 6-3-2018.
 *
 * @author Stefan Breetveld
 */

public class TopCenteredImageView extends android.support.v7.widget.AppCompatImageView {

    public TopCenteredImageView(Context context) {
        super(context);
        init();
    }

    public TopCenteredImageView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public TopCenteredImageView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        super.setScaleType(ScaleType.MATRIX);
    }

    @Override
    public ScaleType getScaleType() {
        return ScaleType.MATRIX;
    }

    /**
     * Ignore all ScaleTypes.
     *
     * @param scaleType ignored.
     */
    @Override
    public void setScaleType(ScaleType scaleType) {
        // Do nothing
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        Matrix matrix = new Matrix();
        matrix.reset();
        Drawable drawable = getDrawable();
        if (drawable == null) return;
        float width = drawable.getIntrinsicWidth();
        float height = drawable.getIntrinsicHeight();
        float xScale = w / width;
        float yScale = h / height;
        float scale;
        if (xScale > yScale) {
            scale = xScale;
        } else {
            float dX = (yScale - xScale) * width;
            matrix.setTranslate(dX, 0);
            scale = yScale;
        }
        matrix.postScale(scale, scale);
        System.out.println(matrix);
        setImageMatrix(matrix);
    }
}

File: app/src/main/java/com/spiderbiggen/randomchampionselector/TopCenteredImageView.kt
package com.spiderbiggen.randomchampionselector

import android.content.Context
import android.graphics.Matrix
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatImageView

/**
 * An [AppCompatImageView] that keeps images horizontally centered and the top at the top of the view
 *
 * @author Stefan Breetveld
 */
class TopCenteredImageView : AppCompatImageView {

    private var viewWidth: Int = 0
    private var viewHeight: Int = 0

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
            super(context, attrs, defStyleAttr)

    init {
        super.setScaleType(ScaleType.MATRIX)
    }

    override fun getScaleType(): ScaleType {
        return ScaleType.MATRIX
    }

    /**
     * Ignore all ScaleTypes.
     *
     * @param scaleType ignored.
     */
    override fun setScaleType(scaleType: ScaleType) {
        // Do nothing
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        this.viewWidth = w
        this.viewHeight = h
        calculateAndSetImageMatrix()
    }

    /**
     * Calculate the [Matrix] for the image so that the top is stuck at the top of the view
     * and the image is horizontally centered. Keep the image as large as possible.
     */
    private fun calculateAndSetImageMatrix() {
        val matrix = Matrix()
        matrix.reset()
        val drawable = drawable ?: return
        val width = drawable.intrinsicWidth.toFloat()
        val height = drawable.intrinsicHeight.toFloat()
        val xScale = this.viewWidth / width
        val yScale = this.viewHeight / height
        val scale: Float
        scale = if (xScale > yScale) {
            xScale
        } else {
            val dX = -0.5f * (yScale - xScale) * width
            matrix.setTranslate(dX, 0f)
            yScale
        }
        matrix.postScale(scale, scale)
        imageMatrix = matrix
    }

    override fun setImageDrawable(drawable: Drawable?) {
        super.setImageDrawable(drawable)
        calculateAndSetImageMatrix()
    }
}
